nbinteract/figure.py

```diff
@@ -0,0 +1,259 @@
+"""
+Methods of the Figure and Plot class to create more complex plots.
+"""
+
+import numpy as np
+import bqplot as bq
+import collections
+import ipywidgets as widgets
+import itertools
+import functools
+import logging
+import toolz.curried as tz
+from IPython.display import display
+from . import util
+from . import plotting
+
+__all__ = ['Hist', 'Bar', 'Scatter', 'Line', 'Plot']
+
+
+def Hist(hist_function, *, options={}):
+    """
+    Public API function to allow users to create a Hist object used to
+    pass into the Plot object,
+    """
+    return {'marker': 'hist', 'x_fn': hist_function, 'options': options}
+
+
+def Bar(x_fn, y_fn, *, options={}):
+    """
+    Public API function to allow users to create a Bar object used to
+    pass into the Plot object.
+    """
+    return {'marker': 'bar', 'x_fn': x_fn, 'y_fn': y_fn, 'options': options}
+
+
+def Scatter(x_fn, y_fn, *, options={}):
+    """
+    Public API function to allow users to create a Scatter object used to
+    pass into the Plot object.
+    """
+    return {
+        'marker': 'scatter',
+        'x_fn': x_fn,
+        'y_fn': y_fn,
+        'options': options
+    }
+
+
+def Line(x_fn, y_fn, options={}):
+    """ Public API function to allow users to create a Line object used to
+    pass into the Plot object.
+    """
+    return {'marker': 'line', 'x_fn': x_fn, 'y_fn': y_fn, 'options': options}
+
+
+@plotting.use_options([
+    'title', 'aspect_ratio', 'animation_duration', 'xlabel', 'ylabel', 'xlim',
+    'ylim', 'bins', 'normalized'
+])
+def _Hist(x_fn, *, options={}):
+    """
+    Internal Hist function that generates the Hist marker and the wrapper
+    functions for the histogram.
+    """
+    params = {
+        'marks': [{
+            'sample': plotting._array_or_placeholder(x_fn),
+            'bins': plotting._get_option('bins'),
+            'normalized': plotting._get_option('normalized'),
+            'scales': (
+                lambda opts: {'sample': opts['x_sc'], 'count': opts['y_sc']}
+            ),
+        }],
+    }
+    fig = options.get('_fig', False)
```

Comment:
Since it looks `_Hist` will only be called with the `_fig` key, we don't need to perform this check.

Same goes for the other internal functions.

Reply:
done!

Reply:
done!